source/common/config/grpc_mux_impl.cc

```diff
@@ -210,6 +220,85 @@ void GrpcMuxImpl::onEstablishmentFailure() {
+    // is already connected to or it is now connected to a different management server instance that
+    // is part of the cluster pointed to by grpc_service. In either case, we should update the
+    // control_plane_identifier in the existing config_source_control_plane_info. Otherwise we
+    // should create a new entry for config_source_control_plane pointing to this grpc_service.
+    for (auto& service_control_plane_info :
+         control_plane_config_dump.service_control_plane_info()) {
+      service_index++;
+      if (service_control_plane_info.xds_service() == xds_service_) {
+        service_config_exists = true;
+        for (auto& config_control_plane :
+             service_control_plane_info.config_source_control_plane()) {
+          config_source_index++;
+          if (message_differencer.Compare(grpc_service_, config_control_plane.grpc_service())) {
+            config_exists = true;
+            break;
+          }
+        }
+        if (config_exists) {
+          break;
+        }
+        config_source_index = -1;
+      }
+    }
+
+    // If service it self does not exist, this is the first time we are seeing it, so add it.
+    if (!service_config_exists) {
+      auto new_service_control_plane_info =
+          control_plane_config_dump.add_service_control_plane_info();
+      new_service_control_plane_info->set_xds_service(xds_service_);
+      auto new_config_source_control_info =
+          new_service_control_plane_info->add_config_source_control_plane();
+      populateControlPlaneIdentifier(new_config_source_control_info,
+                                     message.control_plane().identifier());
+      new_config_source_control_info->mutable_grpc_service()->MergeFrom(grpc_service_);
+    } else {
+      // If service exists, check if the config source exists, if it is theere update it else create
+      // a new config source.
+      auto existing_service_config =
+          control_plane_config_dump.mutable_service_control_plane_info(service_index);
+      if (!config_exists) {
+        auto new_config_source_control_info =
+            existing_service_config->add_config_source_control_plane();
+        populateControlPlaneIdentifier(new_config_source_control_info,
+                                       message.control_plane().identifier());
+      } else {
+        // update the control plane identifier of existing grpc_service.
+        auto existing_config_source_control_info =
+            existing_service_config->mutable_config_source_control_plane(config_source_index);
+        populateControlPlaneIdentifier(existing_config_source_control_info,
+                                       message.control_plane().identifier());
+      }
+    }
+    config_tracker_.addOrUpdateManagedConfig(
+        xds_service_,
+        std::make_shared<envoy::admin::v2alpha::ControlPlaneConfigDump>(control_plane_config_dump));
+  }
+}
```

Comment:
This method seems pretty complicated for what it does. Could you use `map` at the proto level to simplify the data model to avoid iterating etc.?

Reply:
Simplified with `map` in the proto and other simplification in the code. PTAL.